File: runguard/backend/policy/engine.py
"""Policy validation engine — validates actions against safety rules."""

from typing import Any

from runguard.backend.models.policy import Policy

MAX_AUTO_APPROVED_PER_INCIDENT = 5


class PolicyEngine:
    """Validates proposed actions against policy constraints."""

    def validate_action(
        self,
        action_name: str,
        policy: Policy,
        namespace: str = "default",
        environment: str = "dev",
        allowed_namespaces: list[str] | None = None,
        iam_permissions: list[str] | None = None,
        auto_approved_count: int = 0,
    ) -> dict[str, Any]:
        """Validate a single action against the policy.

        Evaluation follows spec rules #1 → #9 (first-match).

        Returns: {"status": "approved"|"requires_approval"|"blocked", "reason": "..."}
        """
        # Rule #1: Forbidden action
        for forbidden in policy.forbidden_actions:
            if forbidden.name == action_name:
                return {
                    "status": "blocked",
                    "reason": f"Forbidden: {forbidden.reason}",
                }

        # Rule #2: Not in allowed list
        allowed_action = None
        for allowed in policy.allowed_actions:
            if allowed.name == action_name:
                allowed_action = allowed
                break

        if allowed_action is None:
            return {
                "status": "blocked",
                "reason": f"Action '{action_name}' not in allowed list",
            }

        # Rule #3: Namespace outside scope
        scope_namespaces = policy.scope.namespaces
        effective_namespaces = allowed_namespaces or scope_namespaces
        if effective_namespaces and namespace not in effective_namespaces:
            return {
                "status": "blocked",
                "reason": f"Scope violation: namespace '{namespace}' not allowed",
            }

        # Rule #4: IAM permissions insufficient
        # AllowedAction doesn't have required_permissions yet; skip if empty
        # In future, each action can declare required IAM permissions
        if iam_permissions is not None:
            pass

        # Rule #5: Production environment
        if environment == "production":
            return {
                "status": "requires_approval",
                "reason": "Production environment requires approval",
            }

        # Rule #6: Blast radius medium/high
        if allowed_action.blast_radius in ("medium", "high"):
            return {
                "status": "requires_approval",
                "reason": (
                    f"Blast radius '{allowed_action.blast_radius}' requires approval"
                ),
            }

        # Rule #7: No rollback path
        if not allowed_action.rollback_path:
            return {
                "status": "requires_approval",
                "reason": "No rollback path defined",
            }

        # Rule #8: Max auto-approved actions reached
        if auto_approved_count >= MAX_AUTO_APPROVED_PER_INCIDENT:
            return {
                "status": "requires_approval",
                "reason": (
                    f"Auto-approve limit reached "
                    f"({MAX_AUTO_APPROVED_PER_INCIDENT} per incident)"
                ),
            }

        # Rule #9: Low risk + rollback + under limit
        return {"status": "approved", "reason": "Low risk action with rollback path"}

    def validate_plan(
        self,
        action_names: list[str],
        policy: Policy,
        namespace: str = "default",
        environment: str = "dev",
        allowed_namespaces: list[str] | None = None,
        iam_permissions: list[str] | None = None,
        auto_approved_count: int = 0,
    ) -> list[dict[str, Any]]:
        """Validate all actions in a remediation plan."""
        results = []
        running_auto_count = auto_approved_count
        for name in action_names:
            result = self.validate_action(
                name,
                policy,
                namespace,
                environment,
                allowed_namespaces,
                iam_permissions,
                running_auto_count,
            )
            results.append(result)
            if result["status"] == "approved":
                running_auto_count += 1
        return results

File: services/operator/internal/policy/engine.go
package policy

import (
	"github.com/phuoctmse/runguard/shared/types"
)

// Classification represents the result of policy validation.
type Classification string

const (
	Approved         Classification = "Approved"
	RequiresApproval Classification = "RequiresApproval"
	Blocked          Classification = "Blocked"
)

// ValidationResult holds the validation result for a single action.
type ValidationResult struct {
	Action         string         `json:"action"`
	Classification Classification `json:"classification"`
	Reason         string         `json:"reason"`
}

// Engine is the unified policy validation engine.
type Engine struct{}

func New() *Engine {
	return &Engine{}
}

// ValidatePlan validates all actions in a remediation plan against the policy.
func (e *Engine) ValidatePlan(
	actions []types.RemediationStep,
	policy types.Policy,
	runbook types.Runbook,
	environment string,
) []ValidationResult {
	results := make([]ValidationResult, 0, len(actions))

	for _, action := range actions {
		result := e.validateSingleAction(action, policy, runbook, environment)
		results = append(results, result)
	}

	return results
}

func (e *Engine) validateSingleAction(
	action types.RemediationStep,
	policy types.Policy,
	runbook types.Runbook,
	environment string,
) ValidationResult {
	proposed := types.ProposedAction{
		Action: action.Action,
		Target: action.Target,
		Risk:   action.Risk,
	}

	// 1. Check forbidden list (highest priority)
	if result := CheckForbidden(proposed, policy); result == Blocked {
		return ValidationResult{
			Action:         action.Name,
			Classification: Blocked,
			Reason:         "action is in forbidden list",
		}
	}

	// 2. Check blast radius (delete action)
	if result := CheckBlastRadiusAction(proposed, policy); result == Blocked {
		return ValidationResult{
			Action:         action.Name,
			Classification: Blocked,
			Reason:         "delete action forbidden by blast radius policy",
		}
	}

	// 3. Check scope
	if result := CheckScope(proposed, environment, policy); result == Blocked {
		return ValidationResult{
			Action:         action.Name,
			Classification: Blocked,
			Reason:         "action outside allowed scope",
		}
	}

	// 4. Check rollback path
	if result := CheckRollback(action, runbook); result == Blocked {
		return ValidationResult{
			Action:         action.Name,
			Classification: Blocked,
			Reason:         "missing rollback path",
		}
	}

	// 5. Check environment rules
	if result := CheckEnvironment(proposed, environment, policy); result == RequiresApproval {
		return ValidationResult{
			Action:         action.Name,
			Classification: RequiresApproval,
			Reason:         "production environment requires approval",
		}
	}

	return ValidationResult{
		Action:         action.Name,
		Classification: Approved,
		Reason:         "all checks passed",
	}
}

// ValidatePlanSafe wraps ValidatePlan with fail-safe: on error, all actions are blocked.
func (e *Engine) ValidatePlanSafe(
	actions []types.RemediationStep,
	policy types.Policy,
	runbook types.Runbook,
	environment string,
) (results []ValidationResult) {
	defer func() {
		if r := recover(); r != nil {
			results = make([]ValidationResult, len(actions))
			for i, action := range actions {
				results[i] = ValidationResult{
					Action:         action.Name,
					Classification: Blocked,
					Reason:         "panic during validation (fail-safe)",
				}
			}
		}
	}()

	if actions == nil {
		return []ValidationResult{{
			Action:         "unknown",
			Classification: Blocked,
			Reason:         "nil actions (fail-safe)",
		}}
	}

	return e.ValidatePlan(actions, policy, runbook, environment)
}

// ValidatePlanWithLimit enforces max auto-approved actions per incident [Req 11.5].
func (e *Engine) ValidatePlanWithLimit(
	actions []types.RemediationStep,
	policy types.Policy,
	runbook types.Runbook,
	environment string,
	maxAutoApproved int,
) []ValidationResult {
	results := e.ValidatePlan(actions, policy, runbook, environment)
	autoApprovedCount := 0

	for i := range results {
		if results[i].Classification == Approved {
			autoApprovedCount++
			if autoApprovedCount > maxAutoApproved {
				results[i].Classification = RequiresApproval
				results[i].Reason = "max auto-approved actions exceeded"
			}
		}
	}

	return results
}
